internal/cmd/backup.go

```diff
@@ -48,7 +71,53 @@ func createBackupFile(filename string) (*os.File, error) {
 	return f, nil
 }
 
-func backupCmdFunc(cmd *cobra.Command, args []string) error {
+var (
+	missingAllowedTypes = regexp.MustCompile(`(\s*)(relation)(.+)(\/\* missing allowed types \*\/)(.*)`)
+	shortRelations      = regexp.MustCompile(`(\s*)relation [a-z][a-z0-9_]:(.+)`)
+)
+
+func filterSchemaDefs(schema, prefix string) (filteredSchema string, err error) {
+	// Remove any invalid relations generated from old, backwards-incompat
+	// Serverless permission systems.
+	schema = string(missingAllowedTypes.ReplaceAll([]byte(schema), []byte("\n/* deleted missing allowed type error */")))
+	schema = string(shortRelations.ReplaceAll([]byte(schema), []byte("\n/* deleted short relation name */")))
+
+	compiledSchema, err := compiler.Compile(compiler.InputSchema{Source: "schema", SchemaString: schema}, compiler.SkipValidation())
+	if err != nil {
+		return "", fmt.Errorf("error reading schema: %w", err)
+	}
+
+	var prefixedDefs []compiler.SchemaDefinition
+	for _, def := range compiledSchema.ObjectDefinitions {
+		if strings.HasPrefix(def.Name, prefix) {
+			prefixedDefs = append(prefixedDefs, def)
+		}
+	}
+
+	for _, def := range compiledSchema.CaveatDefinitions {
+		if strings.HasPrefix(def.Name, prefix) {
+			prefixedDefs = append(prefixedDefs, def)
+		}
+	}
+
+	if len(prefixedDefs) == 0 {
+		return "", errors.New("filtered all definitions from schema")
+	}
+
+	filteredSchema, _, err = generator.GenerateSchema(prefixedDefs)
+	if err != nil {
+		return "", fmt.Errorf("error generating filtered schema: %w", err)
+	}
+	return
+}
+
+func hasRelPrefix(rel *v1.Relationship, prefix string) bool {
+	// Skip any relationships without the prefix on either side.
+	return strings.HasPrefix(rel.Resource.ObjectType, prefix) ||
```

Comment:
I wonder if we want to do this for either side, or if we only want to include relationships that originate on the prefix side...

Reply:
Either way you need the definition, right? I wonder if there's something from SpiceDB's reachability analysis that could be used here for maximum correctness.

Reply:
I used the typesystem to guarantee that our generated schema is comprehensive, this means that if there was any non-prefixed dependencies, we would throw a failure.

We can circle back later and support for discovering the dependencies.